experimental/provider_static/src/lib.rs

```diff
@@ -0,0 +1,78 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use icu_provider::prelude::*;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::fmt::Debug;
+
+const STATIC_STR_DATA: &str = include_str!(concat!(env!("OUT_DIR"), "/static_data.json"));
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(tag = "ty", content = "contents")]
+enum StaticFileOrDir {
+    File(String),
+    Dir(Directory),
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct Directory(Box<HashMap<String, StaticFileOrDir>>);
+
+/// A data provider loading data statically baked in to the binary. Useful for testing in situations
+/// where setting up a filesystem is tricky (e.g. WASM).
+///
+/// This should probably not be used in production code since it bloats the binary.
+pub struct StaticDataProvider {
+    json: Directory,
+}
+
+impl StaticDataProvider {
+    pub fn new() -> Self {
+        StaticDataProvider {
+            json: serde_json::from_str(&STATIC_STR_DATA).unwrap(),
+        }
+    }
+}
+
+impl Default for StaticDataProvider {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<'d, T> DataProvider<'d, T> for StaticDataProvider
+where
+    T: serde::Deserialize<'d> + serde::Serialize + Clone + Debug + 'd,
+{
+    fn load_payload(&self, req: &DataRequest) -> Result<DataResponse<'d, T>, DataError> {
+        let components = req.resource_path.key.get_components();
+        let mut dir = &self.json;
+        let mut file: Option<&str> = None;
+        for component in components
+            .iter()
+            .chain(req.resource_path.options.get_components().iter())
+        {
+            if file.is_some() {
+                return Err(DataError::UnsupportedResourceKey(req.resource_path.key));
+            }
+            let fod = dir
+                .0
+                .get(component)
+                .ok_or(DataError::UnsupportedResourceKey(req.resource_path.key))?;
+            match fod {
+                StaticFileOrDir::Dir(ref d) => dir = d,
+                StaticFileOrDir::File(ref f) => file = Some(f),
+            }
+        }
+        let file = file.ok_or(DataError::UnsupportedResourceKey(req.resource_path.key))?;
```

Comment:
Issue, non-blocking: we should return the correct version of DataError: UnsupportedResourceKey if the resource key is not available, and UnsupportedResourceOptions if the key is available but not for the right locale.